Fath/SIPC Encoding.js

```diff
@@ -0,0 +1,659 @@
+  * @param {string} data Raw input string
+  * @returns {string} Raw MD5 string
+  */
+  function rstrHMACMD5(key, data) {
+    var i
+    var bkey = rstr2binl(key)
+    var ipad = []
+    var opad = []
+    var hash
+    ipad[15] = opad[15] = undefined
+    if (bkey.length > 16) {
+      bkey = binlMD5(bkey, key.length * 8)
+    }
+    for (i = 0; i < 16; i += 1) {
+      ipad[i] = bkey[i] ^ 0x36363636
+      opad[i] = bkey[i] ^ 0x5c5c5c5c
+    }
+    hash = binlMD5(ipad.concat(rstr2binl(data)), 512 + data.length * 8)
+    return binl2rstr(binlMD5(opad.concat(hash), 512 + 128))
+  }
+
+  /**
+  * Convert a raw string to a hex string
+  *
+  * @param {string} input Raw input string
+  * @returns {string} Hex encoded string
+  */
+  function rstr2hex(input) {
+    var hexTab = '0123456789abcdef'
+    var output = ''
+    var x
+    var i
+    for (i = 0; i < input.length; i += 1) {
+      x = input.charCodeAt(i)
+      output += hexTab.charAt((x >>> 4) & 0x0f) + hexTab.charAt(x & 0x0f)
+    }
+    return output
+  }
+
+  /**
+  * Encode a string as UTF-8
+  *
+  * @param {string} input Input string
+  * @returns {string} UTF8 string
+  */
+  function str2rstrUTF8(input) {
+    return unescape(encodeURIComponent(input))
+  }
+
+  /**
+  * Encodes input string as raw MD5 string
+  *
+  * @param {string} s Input string
+  * @returns {string} Raw MD5 string
+  */
+  function rawMD5(s) {
+    return rstrMD5(str2rstrUTF8(s))
+  }
+  /**
+  * Encodes input string as Hex encoded string
+  *
+  * @param {string} s Input string
+  * @returns {string} Hex encoded string
+  */
+  function hexMD5(s) {
+    return rstr2hex(rawMD5(s))
+  }
+  /**
+  * Calculates the raw HMAC-MD5 for the given key and data
+  *
+  * @param {string} k HMAC key
+  * @param {string} d Input string
+  * @returns {string} Raw MD5 string
+  */
+  function rawHMACMD5(k, d) {
+    return rstrHMACMD5(str2rstrUTF8(k), str2rstrUTF8(d))
+  }
+  /**
+  * Calculates the Hex encoded HMAC-MD5 for the given key and data
+  *
+  * @param {string} k HMAC key
+  * @param {string} d Input string
+  * @returns {string} Raw MD5 string
+  */
+  function hexHMACMD5(k, d) {
+    return rstr2hex(rawHMACMD5(k, d))
+  }
+
+  /**
+  * Calculates MD5 value for a given string.
+  * If a key is provided, calculates the HMAC-MD5 value.
+  * Returns a Hex encoded string unless the raw argument is given.
+  *
+  * @param {string} string Input string
+  * @param {string} [key] HMAC key
+  * @param {boolean} [raw] Raw output switch
+  * @returns {string} MD5 output
+  */
+  function md5(string, key, raw) {
+    if (!key) {
+      if (!raw) {
+        return hexMD5(string)
+      }
+      return rawMD5(string)
+    }
+    if (!raw) {
+      return hexHMACMD5(key, string)
+    }
+    return rawHMACMD5(key, string)
+  }
+
+  return md5;
+})();
+/* eslint-enable */
+
+class SIPC_Encoding {
+  constructor(runtime) {
+    this.runtime = runtime
+  })
+}
+
+formatMessage(id) {
+  return this._formatMessage({
```

Comment:
this._formatMessage maybe undefined?

Reply:
_formatMessage is for localization.
if you cant do translate, that's fine, I can help to do that.

Reply:
yes, thank you